Tasks/Common/azure-blobstorage-downloader/azureBlobStorageProvider.ts

```diff
@@ -0,0 +1,139 @@
+import azureStorage = require('azure-storage');
+import models = require('item-level-downloader/Models');
+import Stream = require("stream");
+
+export class AzureBlobProvider implements models.IArtifactProvider {
+    constructor(storageAccount: string, container: string, accessKey: string, prefixFolderPath?: string) {
+        this._storageAccount = storageAccount;
+        this._accessKey = accessKey;
+        this._container = container;
+        this._prefixFolderPath = prefixFolderPath;
+        this._blobSvc = azureStorage.createBlobService(this._storageAccount, this._accessKey);
+        this._maxListSize = 100;
+    }
+
+    public putArtifactItem(item: models.ArtifactItem, readStream: Stream.Readable): Promise<models.ArtifactItem> {
+        return new Promise(async (resolve, reject) => {
+            var newArtifactItem: models.ArtifactItem = models.ArtifactItem.clone(item);
+            await this._ensureContainerExistence();
+
+            var self = this;
+            console.log("Uploading '%s'", item.path);
+            var blobPath = this._prefixFolderPath ? this._prefixFolderPath + "/" + item.path : item.path;
+
+            var writeStream = this._blobSvc.createWriteStreamToBlockBlob(this._container, blobPath, null, function (error, result, response) {
+                if (error) {
+                    console.log("Failed to create blob " + blobPath + ". Error: " + error.message);
+                    reject(error);
+                } else {
+                    var blobUrl = self._blobSvc.getUrl(self._container, blobPath);
+                    console.log("Created blob for item " + item.path + ". Blob uri: " + blobUrl);
+                    newArtifactItem.metadata["downloadUrl"] = blobUrl;
+                    resolve(newArtifactItem);
+                }
+            });
+
+            readStream.pipe(writeStream);
+            writeStream.on("error",
+                (error) => {
+                    reject(error);
+                });
+            readStream.on("error",
+                (error) => {
+                    reject(error);
+                });
+        });
+    }
+
+    public getRootItems(): Promise<models.ArtifactItem[]> {
+        return new Promise(async (resolve, reject) => {
+            var items: models.ArtifactItem[] = await this._getItems(this._container)
+            resolve(items);
+        });
+    }
+
+    public getArtifactItems(artifactItem: models.ArtifactItem): Promise<models.ArtifactItem[]> {
+        return new Promise(async (resolve, reject) => {
+            var items: models.ArtifactItem[] = await this._getItems(this._container, artifactItem.path);
+            resolve(items);
+        });
+    }
+
+    public getArtifactItem(artifactItem: models.ArtifactItem): Promise<Stream.Readable> {
+        return new Promise(async (resolve, reject) => {
+            var readStream: Stream.Readable = this._blobSvc.createReadStream(this._container, artifactItem.path, (error, result) => {
```

Comment:
what is result here. From the signature it would seem that result is the stream.

Reply:
result is the blob object but doesn't contain stream.